### PR TITLE
Serve resources in chunks

### PR DESCRIPTION
**Pull request by staticfloat**
This avoids large amounts of memory consumption when serving large files
such as CUDA artifacts.  It should also reduce TTFB significantly.

### CODE REVIEW

File: src/resource.jl [hunk 1]
```diff
@@ -244,13 +244,21 @@ function download(server::String, resource::String, path::String)
...
-    # Open the path, write it out directly to the HTTP stream
-    open(io -> write(http, read(io, String)), path)
+    # Open the path, write it out directly to the HTTP stream in chunks
+    open(path, "r") do io
+        while !eof(io)
+            # We set `all` to `false` so that we aren't artificially slowing down time to first byte;
+            # if the disk is a little busy, we're happy to start transmitting a smaller chunk of data,
+            # as it is very likely by the time the first small chunk is done transmitting we'll be ready
+            # with a new chunk of data from disk, as the disk is, in general, faster than the network.
+            write(http, read(io, chunksize, all=false))
```
Comment (StefanKarpinski) [inline src/resource.jl]: I wonder if it's better to allocate a buffer and reuse it. cc @Keno

Comment (Keno) [reply src/resource.jl]: Yes, that would be better, unless you plan to do something fancy with parallel streams.

Comment (staticfloat) [reply src/resource.jl]: Death to all allocations.

File: src/resource.jl [hunk 2]
```diff
@@ -244,13 +244,26 @@ function download(server::String, resource::String, path::String)
...
     HTTP.setheader(http, "Content-Length" => string(filesize(path)))
     # We assume that everything we send is gzip-compressed (since they're all tarballs)
     HTTP.setheader(http, "Content-Encoding" => "gzip")
     startwrite(http)
 
-    # Open the path, write it out directly to the HTTP stream
-    open(io -> write(http, read(io, String)), path)
+    # Open the path, write it out directly to the HTTP stream in chunks
+    chunkbuff = Array{UInt8}(undef, chunksize)
+    open(path, "r") do io
```
Comment (StefanKarpinski) [inline src/resource.jl]:
```suggestion
    open(path) do io
```
Open for read is the default anyway.

**Review by StefanKarpinski — APPROVED** [review]

**Pull request merged**